src/wally-connector.ts

```diff
@@ -17,54 +16,100 @@ import {
+  private redirectToCurrentLocation: boolean;
+  private redirectUrl: string | undefined;
   private verbose: boolean;
-  private rejectLogin: (() => void) | null;
+
+  // Internal State
+  private didHandleRedirect = false;
+  private emitterCallbacks: Partial<Record<string, Array<(a?: any) => void>>> =
+    {};
+  private isLoggingIn = false;
+  private worker: SharedWorker | null = null;
+  private workerCallbacks: Partial<Record<WorkerMessage, Array<() => void>>> =
+    {};
 
   constructor({
     clientId,
-    isDevelopment,
     devUrl,
-    token,
+    disableRedirectClose,
+    disableSharedWorker,
+    isDevelopment,
+    disableLoginOnRequest,
+    onTokenFetched,
+    redirectToCurrentLocation,
+    redirectURL,
     verbose,
   }: WallyConnectorOptions) {
     this.clientId = clientId;
+    this.disableRedirectClose = !!disableRedirectClose;
     this.host = (isDevelopment && devUrl) || APP_ROOT;
-    this.selectedAddress = null;
     this.isDevelopment = !!isDevelopment;
-    this.didHandleRedirect = false;
+    this.disableLoginOnRequest = disableLoginOnRequest;
+    this.onTokenFetched = onTokenFetched;
+    this.redirectToCurrentLocation = !!redirectToCurrentLocation;
+    this.redirectUrl = redirectURL;
     this.verbose = !!verbose;
-    this.rejectLogin = null;
 
-    // todo - make path configurable, node_modules maybe?
-    this.worker = SharedWorker ? new SharedWorker('/sdk/worker.js') : null;
-    this.connectToSharedWorker();
-    this.workerCallbacks = {};
+    if (!disableSharedWorker) {
+      // TODO: FIXME!!!!!
+      this.worker = SharedWorker ? new SharedWorker('/sdk/worker.js') : null;
+      this.connectToSharedWorker();
+    }
+  }
 
-    if (token) {
-      if (!isDevelopment) {
-        console.error('Token may only be used in development mode.');
+  public finishLogin = (address: string): void => {
+    if (!this.isLoggingIn) {
+      return;
+    }
+    this.isLoggingIn = false;
+    this.emit('accountsChanged', [address]);
+    this.emit('connected', {});
+  };
+
+  public on(name: string, cb: (a?: any) => void): void {
+    if (!this.emitterCallbacks[name]) {
+      this.emitterCallbacks[name] = [];
+    }
+    this.emitterCallbacks[name]?.push(cb);
+  }
+
+  public addListener(name: string, cb: (a?: any) => void): void {
+    this.on(name, cb);
+  }
+
+  public removeListener(name: string, fn: any): void {
+    const idx = this.emitterCallbacks[name]?.indexOf(fn);
+    if (idx && idx > -1) {
+      this.emitterCallbacks[name]?.splice(idx, 1);
+    }
+  }
+
+  public removeAllListeners(name: string) {
+    this.emitterCallbacks[name] = [];
+  }
+
+  private emit(message: string, value: any): void {
+    this.emitterCallbacks[message]?.forEach((cb) => {
+      if (message === 'accountsChanged') {
+        cb(value);
       } else {
-        if (verbose) {
-          console.log('Setting auth token');
-        }
-        this.setAuthToken(token);
+        cb();
```

Comment:
If value is optional or `undefined`, you can just simplify with a single `cb(value)`.

Reply:
the important part is that we _don't_ want to pass anything when emitting messages other than `accountsChanged` - I'm not entirely sure which messages an application would use, and there isn't really a spec, so I'd rather err on the side of no value rather than an incorrect or misinterpreted value